readme.md
## wheredat [![Build Status](https://jenkins.adc4gis.com/buildStatus/icon?job=wheredat)](https://jenkins.adc4gis.com/view/Open%20Source/job/wheredat/)

A simple, easy to use, geocoding interface for web applications

![WhereDAT](https://raw.github.com/applieddataconsultants/wheredat/master/wheredat.png)

### What is it?

wheredat provides a geocoding and reverse geocoding interface (UI) through
[Bing](http://www.microsoft.com/maps/developers/web.aspx) or [MapQuest](https://developer.mapquest.com/) maps.  You setup an
instance of the application and then embed it in an iframe on any other
applications that use it.  The client app passes an address or lonlat it needs
geocoded and wheredat builds a map showing the location found allowing the user
to move the point on the map to further refine the location.  All this
information is passed back to the client app so it can be used or stored there.

Another one of its goals is to be extremely light weight and quick to load.

### Setup

Built to be served with [node.js](http://nodejs.org/) although you can serve the files up statically as well.

For node do:

```sh
sudo npm install -g wheredat
wheredat [port]
```

### Request Params

- *lat* - latitude to reverse geocode
- *lon* - longitude to reverse geocode
- *address* - address to geocode
- *country* - country to geocode
- *city* - city to geocode **(MapQuest Only)**
- *type* - map type
  - *bing*
    - road
    - aerial
    - aerialwithlabels (default)
  - *mapquest*
    - road
    - satellite
    - hybrid (default)
- *freeze* - lock map, no dragging, zooming, panning
  - true
  - false (default)
- *key* - Bing/MapQuest API key **(required)**
- *layerSwitcher* - Enable layer switcher
  - true
  - false (default)
- *service* - change map service to use mapquest
  - mapquest
  - bing (default)

### Message Data

wheredat uses [window.postMessage](https://developer.mozilla.org/en/DOM/window.postMessage)
for cross domain communication.  However, this is not supported in IE &lt;= 7.
If your application support legacy browsers, wheredat also calls a
function named `wheredat_geocode` on the parent window.  However for this to
work you will need to proxy your wheredat server.

The data returned will look like:

```js
{
   address: "58TH Ave, Chippewa Falls, WI 54729",
   bounds: [ 44.922998091755495, -91.2955657722567, 44.93072352689685, -91.281018090048 ],
   lat: 44.92686080932617,
   lon: -91.28829193115234,
   _mapquestObj: /* Original MapQuest Geocode Object (If using MapQuest) */
   _bingObj: /* Original Bing Geocode Object (If using Bing) */
}
```

Or if unable to geocode:

```js
{
   error: 'wheredat was unable to geocode'
}
```

This data will return whenever a geocode or reverse geocode happens (e.g.
whenever the point is moved or when the map is initially loaded).

### Example

See [the demo page](http://wheredat.adc4gis.com/example.html) for an example of the Bing usage in a client side application.

See [the demo page](http://wheredat.adc4gis.com/mapquest-example.html) for an example of the MapQuest usage in a client side application.

## License

(The MIT License)

Copyright (c) 2014 Applied Data Consultants

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md

0.5.3 / 2015-07-14
==================

  * fix; mapquest geocode pound sign

0.5.2 / 2015-02-13
==================

 * mod; revert urls

0.5.1 / 2015-02-12
==================

  * mod; update mapquest urls
  * add; linting

0.5.0 / 2014-04-03
==================

 * add; layer switcher

0.4.5 / 2014-04-01
==================

 * update; mapquest library versions

0.4.4 / 2014-03-25
==================

 * fix; address line should be called street for mq

0.4.3 / 2014-03-25
==================

 * update; use country and city on geocoder
 * update; license
 * fix; potential undefined country code for bing

0.4.2 / 2014-03-18
==================

 * update; check for https for mapquest [jhays10]

0.4.1 / 2014-03-10
==================

 * fix; resend message for mapquest [jhays10]

0.4.0 / 2014-03-07
==================

 * add; mapquest maps [jhays10]

0.3.0 / 2012-09-28 
==================

  * updated; stringify data object before sending it (breaking change!)

0.2.4 / 2012-05-15 
==================

  * updated; issue with Makefile
  * added; debug option

0.2.3 / 2012-05-15 
==================

  * updated; removed powered by, added debug

0.2.2 / 2012-05-09 
==================

  * updated; default port
  * added; geocodeLat, geocodeLon for geocode match

0.2.1 / 2012-04-24 
==================

  * updated; send a response when geocode failed
  * fix postmessage issue

0.2.0 / 2012-03-01 
==================

  * added; 'freeze' option

0.1.0 / 2012-02-24 
==================

  * Initial release


app.js
"use strict"
var http = require('http')
var fs = require('fs')

process.chdir(__dirname)

function makeIndex() {
   var index = fs.readFileSync('index.html').toString()
      .replace("<link href='lib.css' rel='stylesheet'>", '<style>' + fs.readFileSync('lib.css').toString() + '</style>')
      .replace("<script src='lib.js'></script>", '<script>' + fs.readFileSync('lib.js').toString() + '</script>')
      .replace("<script src='wheredat.js'></script>", '<script>' + fs.readFileSync('wheredat.js').toString() + '</script>')
   return new Buffer(index)
}

var index = makeIndex()
var example = fs.readFileSync('example.html')

var icons = {
   '/img/marker.png': fs.readFileSync('img/marker.png'),
   '/img/marker-shadow.png': fs.readFileSync('img/marker-shadow.png'),
   '/img/popup-close.png': fs.readFileSync('img/popup-close.png'),
   '/img/zoom-in.png': fs.readFileSync('img/zoom-in.png'),
   '/img/zoom-out.png': fs.readFileSync('img/zoom-out.png')
}

var iecss = fs.readFileSync('lib-ie.css')

http.createServer(function (req, res) {
   if (req.url.match(/img/)) {
      res.writeHead(200, { 'Content-Type': 'image/png' })
      res.end(icons[req.url])
   }
   else if (req.url === '/lib-ie.css') {
      res.writeHead(200, { 'Content-Type': 'text/css' })
      res.end(iecss)
   }
   else {
      res.writeHead(200, { 'Content-Type': 'text/html' })
      if (req.url === '/example.html')
         res.end(example)
      else
         res.end(index)
   }
}).listen(8002)


geocad.js
!function _geocad(){
   var BING_LOC_URL = 'https://dev.virtualearth.net/REST/v1/Locations/'

   function param (name) {
      name = name.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]")
      var regexS = "[\\?&]" + name + "=([^&#]*)"
      var regex = new RegExp(regexS)
      var results = regex.exec(window.location.href)
      return results != null ? decodeURIComponent(results[1].replace( /\+/g, " ")) : ''
   }

   var BING_KEY = param('key')

   var G = {
      geocode: function (address) { jsonp('?countryCode=US&q='+address) },
      reverseGeocode: function (lat, lng) { jsonp(lat + ',' + lng) }
   }

   var Icon = L.Icon.extend({
      iconUrl: '/img/marker.png',
      shadowUrl: '/img/marker-shadow.png',
      iconSize: new L.Point(25, 41),
      shadowSize: new L.Point(41, 41),
      iconAnchor: new L.Point(13, 41),
      popupAnchor: new L.Point(0, -33)
   })

   var lat = Number(param('lat'))
   var lon = Number(param('lon'))
   var address = param('address')
   var type = param('type')
   var marker = null
   var addressEl = document.getElementById('address')

   var tsjson = +new Date()
   function jsonp (url) {
      var script = document.createElement('script')
      var q = /\?/.test(url) ? '&' : '?'
      script.src = BING_LOC_URL + url + q + 'key=' + BING_KEY + '&jsonp=_geocad_res&_=' + (tsjson++)
      document.body.appendChild(script)
   }

   var bing = new L.TileLayer.Bing(BING_KEY, 'aerial')

   var map = new L.Map('map', {
      minZoom: 0,
      maxZoom: 21,
      layers: [bing]
   })

   var sendMessage = function (lat, lng) {
      if (!parent) return
      var data = { geocode: { lat: lat, lng: lng } }
      if (parent.postMessage)
         parent.postMessage(data,'*')
      if (parent.geocad_geocode) {
         parent.geocad_geocode(data)
      }
   }

   function createMarker (coords) {
      var markerLocation = new L.LatLng(coords[0], coords[1])
      marker = new L.Marker(markerLocation, { draggable: true, icon: new Icon() })
      marker.on('dragend', handleDrag)
      map.addLayer(marker)
      map.setView(markerLocation, 18)
      sendMessage(coords[0], coords[1])
   }

   function handleDrag (e) {
      var latlng = e.target.getLatLng()
      G.reverseGeocode(latlng.lat,latlng.lng)
      sendMessage(latlng.lat, latlng.lng)
   }

   if (address)
      G.geocode(address)
   else {
      G.reverseGeocode(lat,lon)
      createMarker([lat,lon])
   }

   window._geocad_res = function (data) {
      try {
         var loc = data.resourceSets[0].resources[0]
         console.log(loc)
         if (!marker) { createMarker(loc.point.coordinates) }
         addressEl.innerHTML = loc.name
      } catch (e) { /* */ }
   }
}()


server.js
#!/usr/bin/env node

var http = require('http')
var fs = require('fs')
var port = process.argv[2] || 3000

process.chdir(__dirname)

function makeIndex() {
   var index = fs.readFileSync('index.html').toString()
      .replace("<link href='leaflet/leaflet.css' rel='stylesheet'>", '<style>' + fs.readFileSync('./leaflet/leaflet.css').toString() + '</style>')
      .replace("<script src='leaflet/leaflet.js'></script>", '<script>' + fs.readFileSync('./leaflet/leaflet.js').toString() + '</script>')
      .replace("<script src='wheredat.js'></script>", '<script>' + fs.readFileSync('wheredat.js').toString() + '</script>')
   return new Buffer(index)
}

var index = makeIndex()
var example = fs.readFileSync('example.html')
var mapquestExample = fs.readFileSync('mapquest-example.html')

var icons = {
   '/img/layers.png': fs.readFileSync('./leaflet/img/layers.png'),
   '/img/marker.png': fs.readFileSync('./leaflet/img/marker.png'),
   '/img/marker-shadow.png': fs.readFileSync('./leaflet/img/marker-shadow.png'),
   '/img/popup-close.png': fs.readFileSync('./leaflet/img/popup-close.png'),
   '/img/zoom-in.png': fs.readFileSync('./leaflet/img/zoom-in.png'),
   '/img/zoom-out.png': fs.readFileSync('./leaflet/img/zoom-out.png')
}

var iecss = fs.readFileSync('./leaflet/leaflet-ie.css')

http.createServer(function (req, res) {
   if (req.url.match(/img/)) {
      res.writeHead(200, { 'Content-Type': 'image/png' })
      res.end(icons[req.url])
   }
   else if (req.url === '/leaflet/leaflet-ie.css') {
      res.writeHead(200, { 'Content-Type': 'text/css' })
      res.end(iecss)
   }
   else {
      res.writeHead(200, { 'Content-Type': 'text/html' })
      if (req.url === '/example.html')
         res.end(example)
      else if (req.url === '/mapquest-example.html')
         res.end(mapquestExample)
      else
         res.end(index)
   }
}).listen(port)


wheredat.js
!function _wheredat(){
   var BING_LOC_URL = 'https://dev.virtualearth.net/REST/v1/Locations/'
   var isMapQuest = false

   function param (name) {
      name = name.replace(/[\[]/, "\\[").replace(/[\]]/, "\\]")
      var regexS = "[\\?&]" + name + "=([^&#]*)"
      var regex = new RegExp(regexS)
      var results = regex.exec(window.location.href)
      return results != null ? decodeURIComponent(results[1].replace( /\+/g, " ")) : ''
   }

   var service = param('service')
   var API_KEY = param('key')

   if (service === "mapquest") {
      document.write('<script src="'+location.protocol+'//www.mapquestapi.com/sdk/leaflet/v1.s/mq-map.js?key='+ API_KEY +'"><\/script>');
      document.write('<script src="'+location.protocol+'//www.mapquestapi.com/sdk/leaflet/v1.s/mq-geocoding.js?key='+ API_KEY +'"><\/script>')

      isMapQuest = true
   } else {
      isMapQuest = false
   }

   var tsjson = +new Date()
   function jsonp (url) {
      var script = document.createElement('script')
      var q = /\?/.test(url) ? '&' : '?'
      script.src = BING_LOC_URL + url + q + 'key=' + API_KEY + '&jsonp=_wheredat_res&_=' + (tsjson++)
      document.body.appendChild(script)
   }

   var G = {
      geocode: function (address, country, city) {
         if (!isMapQuest) jsonp('?countryCode='+(country || 'US')+'&q='+address)
         else {
            address = address.replace(new RegExp('#', 'g'), 'no.')
            if (!country || !city) MQ.geocode().search(address).on('success', function(e) { window._wheredat_res(e.result.best) })
            else MQ.geocode().search({ country: country, street: address, city: city }).on('success', function (e) { window._wheredat_res(e.result.best) })
         }
      },
      reverseGeocode: function (lat, lng) {
         if (!isMapQuest) jsonp(lat + ',' + lng)
         else MQ.geocode().reverse({lat:lat,lng:lng}).on('success', function (e) { window._wheredat_res(e.result.best) })
      }
   }

   var Icon = L.icon({
      iconUrl: '/img/marker.png',
      shadowUrl: '/img/marker-shadow.png',
      iconSize: new L.Point(25, 41),
      shadowSize: new L.Point(41, 41),
      iconAnchor: new L.Point(13, 41),
      popupAnchor: new L.Point(0, -33)
   })

   var lat = Number(param('lat'))
   var lon = Number(param('lon'))

   var address = param('address')
   var country = param('country')
   var city = param('city')

   var freeze = param('freeze') === 'true' ? true : false
   var type = param('type')
   var layerSwitcher = param('layerSwitcher') === 'true' ? true : false
   var debug = param('debug') === 'true' ? true : false

   var marker = null
   var addressEl = null

   var layer = null
   var maxZoom = null

   var sameDomain = null
   var map = null
   var lastLatLng = {}
   var baseMaps = null

   var sendMessage = function (loc) {
      if (!parent) return

      var data
      if (loc) {
         data = {
            lat: lastLatLng.lat || (!isMapQuest ? loc.point.coordinates[0] : loc.latlng.lat),
            lon: lastLatLng.lng || (!isMapQuest ? loc.point.coordinates[1] : loc.latlng.lng),
            geocodeLat: !isMapQuest ? loc.point.coordinates[0] : loc.latlng.lat,
            geocodeLon: !isMapQuest ? loc.point.coordinates[1] : loc.latlng.lng,
            address: !isMapQuest ? loc.name : buildAddressStr(loc),
            bounds: !isMapQuest ? loc.bbox : [loc.latlng.lat, loc.latlng.lng, loc.latlng.lat, loc.latlng.lng]
         }
         !isMapQuest ? data._bingObj = loc : data._mapquestObj = loc
      } else
         data = { error: 'wheredat was unable to geocode' }

      if (debug) console.log(data)

      if (parent.postMessage) parent.postMessage(JSON.stringify(data),'*')

      if (sameDomain == null) sameDomain = !!(parent.wheredat_geocode)
      if (sameDomain) parent.wheredat_geocode(JSON.stringify(data))
   }

   function createMarker (coords) {
      var markerLocation = new L.LatLng(coords[0], coords[1])
      marker = new L.Marker(markerLocation, { draggable: freeze ? false : true, icon: Icon })
      if (!freeze) marker.on('dragend', handleDrag)
      map.addLayer(marker)
      map.setView(markerLocation, maxZoom)
   }

   function handleDrag (e) {
      lastLatLng = e.target.getLatLng()
      G.reverseGeocode(lastLatLng.lat,lastLatLng.lng)
   }

   alReady(function() {
      if (isMapQuest) {
         var mapLayer = MQ.mapLayer()
         var hybridLayer = MQ.hybridLayer()
         var satelliteLayer = MQ.satelliteLayer()

         switch (type) {
            case 'hybrid'    : layer = hybridLayer; break
            case 'satellite' : layer = satelliteLayer; break
            case 'road'      : layer = mapLayer; break
            default          : layer = hybridLayer;
         }

         maxZoom = 17
         baseMaps = { "Road": mapLayer, "Hybrid": hybridLayer, "Satellite": satelliteLayer }
      } else {
         var bingRoad = new L.BingLayer(API_KEY, { type: 'road' })
         var bingAerial = new L.BingLayer(API_KEY, { type: 'aerial' })
         var bingAerialLabels = new L.BingLayer(API_KEY, { type: 'aerialwithlabels' })

         switch (type) {
            case 'road'             : layer = bingRoad; break
            case 'aerial'           : layer = bingAerial; break
            case 'aerialwithlabels' : layer = bingAerialLabels; break
            default                 : layer = bingAerialLabels
         }

         maxZoom = 18
         baseMaps = { "Road": bingRoad, "Hybrid": bingAerialLabels, "Satellite": bingAerial }
      }

      addressEl = document.getElementById('address')
      var opt = { minZoom: 0, maxZoom: maxZoom, layers: [layer] }
      if (freeze) {
         opt.touchZoom = false
         opt.scrollWheelZoom = false
         opt.doubleClickZoom = false
         opt.dragging = false
         opt.zoomControl = false
      }
      map = new L.Map('map', opt)
      if (layerSwitcher) map.addControl(L.control.layers(baseMaps))

      if (address || city || country) G.geocode(address, country, city)
      else {
         G.reverseGeocode(lat,lon)
         createMarker([lat,lon])
      }
   })

   window._wheredat_res = function (data) {
      try {
         var loc = !isMapQuest ? data.resourceSets[0].resources[0] : data
         if (!loc) return sendMessage(null)

         if (!isMapQuest) {
            if (!marker) { createMarker(loc.point.coordinates) }
            addressEl.innerHTML = loc.name
         } else {
            if (!marker) { createMarker([loc.latlng.lat, loc.latlng.lng]) }
            addressEl.innerHTML = loc.adminArea5 !== "" && loc.adminArea3 !== "" ? buildAddressStr(loc) : ""
         }

         sendMessage(loc)
      } catch (e) {}
   }

   function buildAddressStr (data) {
      return data.street !== "" ? data.street + ", " + data.adminArea5 + ", " + data.adminArea3 + " " + data.postalCode
                                : data.adminArea5 + ", " + data.adminArea3 + " " + data.postalCode
   }
}()
